### PR TITLE
Fix race condition in SharedTable

### PR DESCRIPTION
**Pull request by mihacooper**

### CODE REVIEW

File: src/cpp/shared-table.cpp [hunk 1]
```diff
@@ -280,7 +280,8 @@ SharedTable SharedTable::luaSetMetatable(const sol::stack_object& tbl, const sol
         stable.ctx_->metatable = GCNull;
     }
 
-    stable.ctx_->metatable = createStoredObject(mt)->gcHandle();
+    const auto mtObj = createStoredObject(mt);
```
Comment (IUdalov) [inline src/cpp/shared-table.cpp:283]: Where is the race?

Comment (mihacooper) [reply src/cpp/shared-table.cpp:283]: returned stored object could be destroyed by GC before we add using `addReference`

Comment (mihacooper) [reply src/cpp/shared-table.cpp]: returned stored object could be destroyed by GC before we add using `addReference`

Comment (mihacooper) [reply src/cpp/shared-table.cpp]: returned stored object could be destroyed by GC before we add using `addReference`

Comment (mihacooper) [reply src/cpp/shared-table.cpp]: returned stored object could be destroyed by GC before we add using `addReference`

Comment (mihacooper) [reply src/cpp/shared-table.cpp]: returned stored object could be destroyed by GC before we add using `addReference`

Comment (mihacooper) [reply src/cpp/shared-table.cpp]: returned stored object could be destroyed by GC before we add it using `addReference`

Comment (mihacooper) [reply src/cpp/shared-table.cpp]: returned stored object could be destroyed by GC before we add it using `addReference`

Comment (mihacooper) [reply src/cpp/shared-table.cpp:283]: it could be destroyed by GC before we save it using `addReference`

**Review by IUdalov — APPROVED** [review]

**Pull request merged**